### DISCUSSION TITLE
useStore Probleme mit nested objects und zustand

### DISCUSSION THREAD

**Discussion by AnH13-6** [neuefische/questions#157] [Web]
Hallo zusammen, 

ich habe das Problem, dass ich text-entries mit einem Datum zu einer Liste hinzufügen will, die wiederrum in einem größeren Objekt von vacations liegen. Dabei ist auch wichtig, dass die entries nur unter der entsprechenden vacation aufrufbar sein sollen. Ich dachte, ich könnte das mit "immer" lösen, aber ich habe die Lösung noch nicht gefunden. Zuletzt hatte ich es noch mal ohne "immer" versucht, da ich dachte, das sei vielleicht einfacher, aber es hat mich auch nicht weitergebracht.  Hauptsächlich hab ich mich bei meinen Versuchen an dieser Seite orientiert: https://beta.reactjs.org/learn/updating-objects-in-state

Hier der Link zu meinem Repo mit den Problemen.
https://github.com/AnH13-6/capstone-project/commit/51536c5fde50c054af88d66afebf2ec464445cfe

Hier das verschachtelte Object in meinem useStore: 
```javascript
import {nanoid} from 'nanoid';
import create from 'zustand';

let vacationStore = set => ({
	vacations: [
		{
			vacationName: 'Italy 2021',
			id: nanoid(),
			entries: [{text: 'Test Entry', id: nanoid(), date: 'date'}],
		},
		{
			vacationName: 'Japan 2015',
			id: nanoid(),
			entries: [],
		},
	],
	addVacation: vacation => {
		set(state => ({
			vacations: [vacation, ...state.vacations],
		}));
	},
	removeVacation: vacationId => {
		set(state => ({
			vacations: state.vacations.filter(vacation => vacation.id !== vacationId),
		}));
	},

	handleEntryAdd: vacation => {
		set(state => ({
			vacations: [{vacation, ...state.vacations.entries, entries: event.target.value}],
		}));
	},
});

const useVacationStore = create(vacationStore);

export default useVacationStore;

**Comment by actyvystom** [comment]
Hallo Anne, grundsätzlich ist `immer` dafür eine gute Lösung. Allerdings brauchst Du in deiner Form zur Aktualisierung deiner `entries` aber auch eine `id` der Vacation, der du dann die neuen `entries` mit `date` und `text` hinzufügen kannst!
Aber der Reihe nach:

In deinem VacationStore-Hook importierst Du zunächst `produce` aus immer: 
```javascript
import produce from 'immer';
```  

und implementierst eine Funktion, um dein Vacation-Objekt zu aktualisieren, z.B.

```javascript
updateVacation: vacation =>
		set(
			produce(draft => {
				const vacationToUpdate = draft.vacations.find(el => el.id === vacation.id);
				vacationToUpdate.entries.push({date: vacation.date, text: vacation.entry});
			})
		)
```

In deiner EntryForm-Komponente kannst du diese Funktion dann folgendermaßen verwenden: 

- Extrahiere die Funktion aus deinem Store-Hook:

```javascript
const {vacations, updateVacation} = useVacationStore();
```
- deklariere eine einfache State-Variable, um deine ausgewählte `id` (z.B. aus einer Select-Box heraus oder - wie wir es **vereinfacht** in der Demo gemacht haben - über einen `<a>-Tag` zu setzen):

```javascript
const [selectedVacationId, setSelectedVacationId] = useState('');
```
Die vereinfachte (aber wenig elegante) Version aus der Demo sah so aus:

```javascript
{vacations.map(vacation => (
<a key={vacation.id} onClick={() => setSelectedVacationId(vacation.id)}>{vacation.vacationName}</a>
))}
``` 


In deiner Form nimmst Du diese State-Variable in einem hidden-Input auf, damit Du beim Submit der Form die `id` in deinen `formValues` verfügbar hast: 

```html
<input type={'hidden'} value={selectedVacationId} name="id" />
``` 

Dann kannst du im `onSubmit`-Handler deiner Form die `updateVacation`-Funktion mit deinen `formValues` aufrufen: 

```javascript
...
updateVacation(formValues);

```
Falls Du noch Fragen dazu hast, sag gerne Bescheid!
